.scripts/ls.py

```diff
@@ -0,0 +1,173 @@
+
+logging.basicConfig(level=logging.DEBUG)
+logger = logging.getLogger(str(pathlib.Path(__file__).relative_to(_REPO_ROOT)))
+
+
+@dataclasses.dataclass
+class _Args:
+    kind: str
+    refs: tuple[str, str | None] | None
+    include_examples: bool
+    include_placeholders: bool
+    only_if_version_changed: bool
+
+    @classmethod
+    def from_cli(cls) -> _Args:
+        parser = argparse.ArgumentParser()
+        parser.add_argument('kind', choices=('packages', 'interfaces'))
+        parser.add_argument('old_ref', nargs='?')
+        parser.add_argument('new_ref', nargs='?')
+        parser.add_argument('--exclude-examples', action='store_true')
+        parser.add_argument('--exclude-placeholders', action='store_true')
+        parser.add_argument('--only-if-version-changed', action='store_true')
+        args = parser.parse_args()
+        return cls(
+            kind=args.kind,
+            refs=(args.old_ref, args.new_ref) if args.old_ref is not None else None,
+            include_examples=not args.exclude_examples,
+            include_placeholders=not args.exclude_placeholders,
+            only_if_version_changed=args.only_if_version_changed,
+        )
+
+
+def _ls(args: _Args) -> list[str]:
+    include: list[str] = []
+    if args.include_examples:
+        include.extend(('.example', '.tutorial'))
+    if args.include_placeholders:
+        include.append('.package')
+    # collect packages or interfaces
+    if args.kind == 'packages':
+        dirs = _packages(include=include)
+    elif args.kind == 'interfaces':
+        dirs = _interfaces(include=include)
+    else:
+        raise ValueError(f'Unknown value for `kind` {args}')
+    # filter based on changes
+    if args.refs:
+        old_ref, new_ref = args.refs
+        dirs = _changed_only(dirs, old_ref=old_ref, new_ref=new_ref)
+        if args.only_if_version_changed and args.kind == 'packages':
+            dirs = _changed_version_only(dirs, old_ref=old_ref, new_ref=new_ref)
+    return [str(p) for p in dirs]
+
+
+def _packages(include: list[str]) -> list[pathlib.Path]:
+    paths: set[pathlib.Path] = set()
+    for root in _REPO_ROOT, _INTERFACES:
+        paths.update(root.glob(r'[a-z]*'))
+        paths.update(root / i for i in include)
+    return sorted(p.relative_to(_REPO_ROOT) for p in paths if (p / 'pyproject.toml').exists())
+
+
+def _interfaces(include: list[str]) -> list[pathlib.Path]:
+    paths = {*_INTERFACES.glob(r'[a-z]*'), *(_INTERFACES / i for i in include)}
+    return sorted(p.relative_to(_REPO_ROOT) for p in paths if (p / 'interface').is_dir())
+
+
+def _changed_only(
+    dirs: list[pathlib.Path], old_ref: str, new_ref: str | None
+) -> list[pathlib.Path]:
+    cmd = ['git', 'diff', '--name-only', old_ref]
+    if new_ref is not None:
+        cmd.append(new_ref)
+    names = subprocess.check_output(cmd, text=True).strip().splitlines()
+    if new_ref is None:  # include untracked files when run w/out explicit new ref for local tests
```

Comment:
Feel free to disregard, but I feel like "w/out" would be better as "without" (and on the line above if it's shortened only for line length reasons).